test/TestPoolCreation.ts

```diff
@@ -0,0 +1,221 @@
+
+describe("HappyTokenPool", () => {
+  let creationParams: HappyPoolParamType; // fill happyTokenPoolDeployed parameters
+  let snapshotId: string;
+  let testTokenADeployed: TestToken;
+  let testTokenBDeployed: TestToken;
+  let testTokenCDeployed: TestToken;
+
+  let happyTokenPoolDeployed: HappyTokenPool;
+  let qualificationTesterDeployed: QLF;
+
+  let signers: Signer[];
+  let creator: Signer;
+  let ito_user: Signer;
+
+  before(async () => {
+    signers = await ethers.getSigners();
+    creator = signers[0];
+    ito_user = signers[1];
+
+    const TestTokenA = await ethers.getContractFactory("TestToken");
+    const TestTokenB = await ethers.getContractFactory("TestToken");
+    const TestTokenC = await ethers.getContractFactory("TestToken");
+    const QualificationTester = await ethers.getContractFactory("QLF");
+
+    const testTokenA = await TestTokenA.deploy(amount, "TestTokenA", "TESTA");
+    const testTokenB = await TestTokenB.deploy(amount, "TestTokenB", "TESTB");
+    const testTokenC = await TestTokenC.deploy(amount, "TestTokenC", "TESTC");
+    const qualificationTester = await QualificationTester.deploy(0);
+
+    testTokenADeployed = (await testTokenA.deployed()) as TestToken;
+    testTokenBDeployed = (await testTokenB.deployed()) as TestToken;
+    testTokenCDeployed = (await testTokenC.deployed()) as TestToken;
+    qualificationTesterDeployed = (await qualificationTester.deployed()) as QLF;
+
+    const HappyTokenPool = await ethers.getContractFactory("HappyTokenPool");
+    const HappyTokenPoolProxy = await upgrades.deployProxy(HappyTokenPool, [base_timestamp], {
+      unsafeAllow: ["delegatecall"],
+    });
+    happyTokenPoolDeployed = new ethers.Contract(
+      HappyTokenPoolProxy.address,
+      itoJsonABI.abi,
+      creator,
+    ) as HappyTokenPool;
+  });
+
+  beforeEach(async () => {
+    snapshotId = await takeSnapshot();
+    creationParams = {
+      hash: ethers.utils.keccak256(ethers.utils.toUtf8Bytes(PASSWORD)),
+      start_time: 0,
+      end_time: 10368000, // duration 120 days
+      message: "Hello From the Outside Hello From the Outside",
+      exchange_addrs: [eth_address, testTokenBDeployed.address, testTokenCDeployed.address],
+      exchange_ratios: [1, 10000, 1, 2000, 4000, 1],
+      lock_time: 12960000, // duration 150 days
+      token_address: testTokenADeployed.address,
+      total_tokens: ethers.utils.parseEther("10000"),
+      limit: ethers.utils.parseEther("1000"),
+      qualification: qualificationTesterDeployed.address,
+    };
+    const nowTimeStamp = Math.floor(new Date().getTime() / 1000);
+    // 120 days
+    creationParams.end_time = nowTimeStamp + 10368000 - base_timestamp;
+    // 120 days
+    creationParams.lock_time = nowTimeStamp + 12960000 - base_timestamp;
+  });
+
+  afterEach(async () => {
+    await revertToSnapShot(snapshotId);
+  });
+
+  describe("constructor()", async () => {
+    it("Should variables be initalized properly", async () => {
```

Comment:
Typo here. `initalized`. You could use cspell to check if there exists any typo in your codes.